fiftyone/utils/multiprocessing.py

```diff
@@ -0,0 +1,572 @@
+        worker_progress = False
+        lock = None
+
+    if return_outputs:
+        batch_count = multiprocessing.Value("i", 0)
+        sample_count = None
+        queue = multiprocessing.Queue()
+    elif progress != False:
+        batch_count = multiprocessing.Value("i", 0)
+        sample_count = multiprocessing.Value("i", 0)
+        queue = None
+    else:
+        batch_count = None
+        sample_count = None
+        queue = None
+
+    pool = ctx.Pool(
+        processes=num_workers,
+        initializer=_init_worker,
+        initargs=(
+            dataset_name,
+            view_stages,
+            pickle.dumps(map_fcn),
+            batch_count,
+            sample_count,
+            queue,
+            save,
+            worker_progress,
+            lock,
+        ),
+    )
+
+    pb = fou.ProgressBar(total=num_samples, progress=progress)
+
+    if reducer is not None:
+        return _do_map_reduce_samples(
+            pool, pb, batches, batch_count, queue, reducer
+        )
+    elif queue is not None:
+        return _do_map_samples(pool, pb, batches, batch_count, queue)
+    else:
+        return _do_update_samples(pool, pb, batches, batch_count, sample_count)
+
+
+def _do_update_samples(pool, pb, batches, batch_count, sample_count):
+    num_batches = len(batches)
+
+    with pool, pb:
+        pool.map_async(_map_batch, batches)
+
+        if batch_count is not None:
+            while batch_count.value < num_batches:
+                pb.set_iteration(sample_count.value)
+                time.sleep(0.01)
+
+            pb.set_iteration(sample_count.value)
+
+        pool.close()
+        pool.join()
+
+
+def _do_map_samples(pool, pb, batches, batch_count, queue):
+    num_batches = len(batches)
+
+    with pool, pb:
+        pool.map_async(_map_batch, batches)
+
+        while True:
+            try:
+                result = queue.get(timeout=0.01)
+                pb.update()
+                yield result
+            except Empty:
+                if batch_count.value >= num_batches:
+                    break
+
+        queue.close()
+        queue.join_thread()
+
+        pool.close()
+        pool.join()
+
+
+def _do_map_reduce_samples(pool, pb, batches, batch_count, queue, reducer):
+    num_batches = len(batches)
+
+    reducer.init()
+
+    with pool, pb:
+        pool.map_async(_map_batch, batches)
+
+        while True:
+            try:
+                result = queue.get(timeout=0.01)
+                pb.update()
+                reducer.add(*result)
+            except Empty:
+                if batch_count.value >= num_batches:
+                    break
+
+        queue.close()
+        queue.join_thread()
+
+        pool.close()
+        pool.join()
+
+    return reducer.finalize()
+
+
+def _map_samples_single(
+    sample_collection,
+    map_fcn,
+    reduce_fcn=None,
+    aggregate_fcn=None,
+    return_outputs=True,
+    save=False,
+    progress=None,
+):
+    if reduce_fcn is not None:
```

Comment:
We probably should raise an error/warning if both `reduce_fnc` and `aggregate_fcn` are passed in